### PR TITLE
[do-not-merge]Ispn 6561/spring boot starters remote

### PR DESCRIPTION
**Pull request by slaskawi**
Created because of documentation reasons...

To be closed in a day or two.

### CODE REVIEW

File: spring-boot/src/main/resources/hotrod-client.properties [hunk 1]
```diff
@@ -0,0 +1 @@
+infinispan.client.hotrod.server_list=127.0.0.1:11222
```
Comment (snicoll) [inline spring-boot/src/main/resources/hotrod-client.properties]: FYI, The canonical format for properties is to use `-` not `_`

File: spring-boot/pom.xml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        <dependency>
+            <groupId>org.infinispan</groupId>
+            <artifactId>infinispan-remote</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.infinispan</groupId>
+            <artifactId>inifinispan-spring-boot-starter</artifactId>
+            <version>${infinispan.starters.version}</version>
+        </dependency>
+        <dependency>
```
Comment (snicoll) [inline spring-boot/pom.xml]: We shouldn't have to do this IMO. Your starter should bring `spring-boot-starter` with it.

File: spring-boot/pom.xml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    <dependencies>
+        <dependency>
+            <groupId>org.infinispan</groupId>
+            <artifactId>infinispan-remote</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.infinispan</groupId>
+            <artifactId>inifinispan-spring-boot-starter</artifactId>
+            <version>${infinispan.starters.version}</version>
```
Comment (snicoll) [inline spring-boot/pom.xml]: Why do you need a version here and not for `infinispan-remote` ?

File: spring-boot/src/main/resources/hotrod-client.properties (hunk 1 above)
Comment (slaskawi) [inline spring-boot/src/main/resources/hotrod-client.properties]: This property file is read by a `RemoteCacheManager`'s configuration [here](https://github.com/infinispan/infinispan-spring-boot/pull/7/files#diff-633ddf90ddf8ca6ccee9e06c7bc964ceR47). It is written in our format, please have a look into [the documentation](http://infinispan.org/docs/stable/user_guide/user_guide.html#configuration_9).

Hot Rod Client is used in Client/Server mode, where a client connects to a remote Server.

File: spring-boot/pom.xml (hunk 2 above)
Comment (slaskawi) [inline spring-boot/pom.xml]: That's a good point, but than we would need to lock the version (to `1.4.0.RELEASE` for the instance). Is that OK?

File: spring-boot/pom.xml (hunk 3 above)
Comment (slaskawi) [inline spring-boot/pom.xml]: `infinispan-remote`'s version is defined in [parent module of this tutorial](https://github.com/infinispan/infinispan-simple-tutorials/blob/master/pom.xml#L19). 

Please have in mind that these tutorials show many different integration points for Infinispan. Spring Boot is one of them (that's why Spring Boot's version is defined in this module and not the parent).

**Comment by slaskawi** [comment]
So it seems we only need to clarify whether or not we should define a compile dependency to `spring-boot-starter`. That's the main discussion point in my opinion. 

The main argument for it is simplicity (it's automatically included in user's classpath). An argument against it is enforcing specific version (what will happen is someone wants to use `1.4.2.RELEASE` and not `1.4.0.RELEASE`). @snicoll WDYT?

File: spring-boot/pom.xml (hunk 3 above)
Comment (snicoll) [inline spring-boot/pom.xml]: IMO your sample should expose how the project will be configured in the target environnement. No promise here and even though we've never done this before, we might provide dependency management for this starter in Spring Boot itself, since we provide dependency management for Infinispan.

File: spring-boot/pom.xml (hunk 2 above)
Comment (snicoll) [inline spring-boot/pom.xml]: It doesn't matter what you specify there. Spring Boot comes with its own `dependencyManagement` per version that is going to override anything provided by 3rd parties. What matters though is which version of Spring Boot you are compatible against. I see no reason to go 1.4 for your current usage. Perhaps you could build against 1.3 and 1.4?

**Comment by snicoll** [comment]
> what will happen is someone wants to use 1.4.2.RELEASE and not 1.4.0.RELEASE

It will work out of the box. Your starter will state "please bring spring boot 1.4.0.RELEASE" and a Spring Boot-based project will have an explicit `dependencyManagement` for that, overriding your decision. Same thing with Gradle.

I disagree that's the only concern though. Our discussion on embedded and remote is very much relevant IMO. If your user has yet another decision to make after adding your starter, something isn't right.

File: spring-boot/src/main/resources/hotrod-client.properties (hunk 1 above)
Comment (slaskawi) [inline spring-boot/src/main/resources/hotrod-client.properties]: In general I think that's a good idea... However we use [a lot of those properties](https://github.com/infinispan/infinispan/blob/master/client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/ConfigurationProperties.java) and duplicating them in the starter adds some additional maintenance effort.

I've created [an issue for implementing this](https://github.com/infinispan/infinispan-spring-boot/issues/8) but I will probably look into it after the initial release.

File: spring-boot/pom.xml (hunk 2 above)
Comment (slaskawi) [inline spring-boot/pom.xml]: Ok, got it. I will modify the starter than.

File: spring-boot/pom.xml (hunk 3 above)
Comment (slaskawi) [inline spring-boot/pom.xml]: I'm not entirely sure if I understand you correctly...

This is a tutorial project which shows how to use Infinispan's features (Spring Boot Starter as mentioned in this example but also how to use queries, C# client etc).

So what exactly is your request here?

**Comment by slaskawi** [comment]
Alrighty... Most of the comments are addressed now. 

In order to use the client/server mode we need to exclude the core and add dependency to Hot Rod client. Also there is no need to define any additional dependencies to Spring Boot artifacts. Everything is pulled transitively.

**Comment by slaskawi** [comment]
@snicoll Just a friendly reminder

**Comment by snicoll** [comment]
Do you have updated samples for both use cases?

**Comment by slaskawi** [comment]
Yes, it's this PR (I force-pushed the changes) and [this one](https://github.com/infinispan/infinispan-simple-tutorials/pull/19).

File: spring-boot/pom.xml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    <dependencies>
+        <dependency>
+            <groupId>org.infinispan</groupId>
+            <artifactId>inifinispan-spring-boot-starter</artifactId>
+            <version>${infinispan.starters.version}</version>
+            <exclusions>
+                <exclusion>
+                    <groupId>org.infinispan</groupId>
+                    <artifactId>infinispan-core</artifactId>
+                </exclusion>
```
Comment (snicoll) [inline spring-boot/pom.xml]: This is  weird. So the starter brings the embedded mode and you have to remove it to add the client-side mode. Everything in Spring Boot works the other way around: the starter brings the client/server support and if you want to embed (which arguably is going to bring more dependencies), then you should add it. 

The naming doesn't help, `infinispan-core` makes me think of something I should always have. 

Is there a reason you are doing this the other way around. Having to exclude something to add the lightweight option makes my eyes bleed.

File: spring-boot/pom.xml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        <version>1.0.0-SNAPSHOT</version>
+        <groupId>org.infinispan.tutorial.simple</groupId>
+        <artifactId>infinispan-simple-tutorials</artifactId>
+    </parent>
+
+    <name>Infinispan Simple Tutorials: Spring Boot</name>
+
+    <properties>
+        <spring-boot.version>1.4.0.RELEASE</spring-boot.version>
+        <infinispan.starters.version>1.0.0-SNAPSHOT</infinispan.starters.version>
```
Comment (snicoll) [inline spring-boot/pom.xml]: Have you considered releasing the starter with infinispan itself? That way the starter is aligned with the infinispan version, no further complex dependency management would be required.

Comment (slaskawi) [inline spring-boot/pom.xml]: Of course! Once we agree upon all the comments, I'll release it.

Comment (snicoll) [inline spring-boot/pom.xml]: Sorry, I mean the same version than infinispan rather than `1.0.0`.

File: spring-boot/pom.xml (hunk 4 above)
Comment (slaskawi) [inline spring-boot/pom.xml]: If I made the Client/Server mode the default, the starter woul not be "self-contained". In other words, with the default setup you will need an Infinispan Server running somewhere and connect to it. 

with Embedded mode as a default, everything works with zero configuration.

So the question is - is it better to have something lightweight (and have a running Server somewhere) or is it better to have something self-contained (that works with no additional configuration). If I understood you correctly, Spring Starters prefer the former?

Comment (snicoll) [inline spring-boot/pom.xml]: It's not about self-contained, it's about what you have to do if you want "more". I think you're asking the wrong question: Spring Boot starter have no "preference". The only reason why it exist is to serve the (infinispan) user and not go in their way. 

If the vast majority of your users use the embedded mode then by all means provide that by default (but you shouldn't have to exclude it to use the client mode). Have you looked at our JMS support? It's the exact same thing and has a lot maturity so maybe [you should look there for inspiration](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jms/artemis/ArtemisConnectionFactoryFactory.java#L95-L105)). So, we look in the classpath: if we have both, we prefer the embedded mode but there is a property the user can set to specify in which mode they want to run (for JMS that's `spring.artemis.embedded.enabled`).

Long story short: it does not matter if the starter is self contained or not. What matters is what the starter should logically do by default. We usually go with client/server because adding the "whole server" by default can be quite involved (but I reckon that caching libraries have a lower footprint). If the vast majority of your users use infinispan in X mode, that's what the starter should do by default.

**Comment by slaskawi** [comment]
@snicoll Ok, updated. The client/server configuration requires no additional dependencies and embedded requires `infinispan-core`.

**Comment by snicoll** [comment]
What happens if you have both and you want to use client/slide?

**Comment by slaskawi** [comment]
Ohhh right, sorry... The time flies to fast for me today...

Let me add those two missing switches and get back to you.

**Comment by slaskawi** [comment]
@snicoll Switches for both embedded and client/server mode are implemented. Now all you need to do is to use any combination of those two:

```
infinispan.remote.enabled=false
infinispan.embedded.enabled=false
```

**Comment by snicoll** [comment]
Sorry to insist but is there any reason to invent another mechanism compared to what I've referenced before? What happens when you set bot those values to true? That's confusing, isn't it?

**Comment by slaskawi** [comment]
I think it actually works as you described and it is based on Artemis starter (see [this part for example](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jms/artemis/ArtemisProperties.java#L124)).

And it's absolutely perfect scenario to have both extensions enabled (or disabled) because both of them are guarded by specific class availability (see [embedded](https://github.com/infinispan/infinispan-spring-boot/pull/7/files#diff-0abd7fb5b7e7a90d4522dd09881f4730R24) and [remote](https://github.com/infinispan/infinispan-spring-boot/pull/7/files#diff-633ddf90ddf8ca6ccee9e06c7bc964ceR25)). So it should work with all combinations.

**Comment by snicoll** [comment]
uh? So it makes sense to have both a remote client and both an embedded server at the same time in the same app? I didn't know that, sorry.

**Comment by slaskawi** [comment]
Yes, it's perfectly OK.

Allrighty... Thanks a lot for a huuuuge amount of help @snicoll. I really appreciate it!

I'll make an `Alpha` release tomorrow and we'll see what the community thinks about it.

**Comment by slaskawi** [comment]
Closing as this was an example mainly for discussion. For more information please have a look at: https://github.com/infinispan/infinispan-simple-tutorials/pull/19
